keps/sig-auth/3299-kms-v2-improvements/README.md

```diff
@@ -0,0 +1,524 @@
+
+```proto
+message EncryptResponse {
+    // The encrypted data.
+    bytes ciphertext = 1;
+    // The KMS key ID used for encryption operations.
+    // This can be used to drive rotation.
+    string current_key_id = 2;
+    // Additional metadata to be stored with the encrypted data.
+    // This metadata can contain the encrypted local KEK that was used to encrypt the DEK.
+    // Stored unencrypted in etcd.
+    map<string, string> metadata = 3;
+}
+```
+
+The `DecryptRequest` passes the same `current_key_id` and `metadata` returned by the previous `EncryptResponse` of this data as its `observed_key_id` and `metadata` for the decryption request.
+
+```proto
+message DecryptRequest {
+    // The data to be decrypted.
+    bytes ciphertext = 2;
+    // UID is a unique identifier for the request.
+    string uid = 3;
+    // The keyID that was provided to the apiserver during encryption.
+    // This represents the KMS KEK that was used to encrypt the data.
+    string observed_key_id = 4;
+    // Additional metadata that was sent by the KMS plugin during encryption.
+    map<string, string> metadata = 5;
+}
+
+message DecryptResponse {
+    // The decrypted data.
+    bytes plaintext = 1;
+    // The KMS key ID used to decrypt the data.
+    string current_key_id = 2;
+    // Additional metadata that was sent by the KMS plugin.
+    map<string, string> metadata = 3;
+}
+
+message EncryptRequest {
+    // The data to be encrypted.
+    bytes plaintext = 2;
+    // UID is a unique identifier for the request.
+    string uid = 3;
+}
+```
+
+In terms of storage, a new structured protobuf format is proposed. Similar to the proto serializer, it will use a magic number to detect when the stored data is in a format that it understands:
+
+```go
+encryptedProtoEncodingPrefix = []byte{'e', 'k', '8', 's', 0}
+```
+
+The last byte represents the encoding style, with 0 meaning that the rest of the byte stream is a proto message of type `EncryptedObject`:
+
+```go
+type EncryptedObject struct {
+    TypeMeta `json:",inline" protobuf:"bytes,1,opt,name=typeMeta"`
+
+    EncryptResponse EncryptResponse `protobuf:"bytes,2,opt,name=encryptResponse"`
+
+    PluginName string `protobuf:"bytes,3,opt,name=pluginName"`
+
+    Ciphertext []byte `protobuf:"bytes,4,opt,name=ciphertext"`
+}
+```
+
+This object simply provides a structured format to store the `EncryptResponse` data with the plugin name and encrypted object data. New fields can easily be added to this format.
+
+To improve health check reliability, the new StatusResponse provides version, healthz information, and can trigger key rotation via storage version status updates.
+
+```proto
+message StatusRequest {}
+
+message StatusResponse {
+    // Version of the KMS plugin API.
+    string version = 1;
+
+    // anything other than "ok" is failing healthz
+    string healthz = 2;
```

Comment:
Is this informational/passthrough to $KUBE_APISERVER/healthz or will the kube-apiserver react in some way if this isn't "ok"? We might want to use conditions here.

Reply:
This is the same behavior as the current healthcheck functionality such that when the kms healthcheck fails, the information will passthrough to apiserver healthz. This change only impacts _how_ the healthcheck is done, not how the data is passed to the apiserver.